File: app/src/main/java/com/cyclecare/app/domain/engine/CyclePredictionEngine.kt
package com.cyclecare.app.domain.engine

import com.cyclecare.app.domain.model.Period
import com.cyclecare.app.domain.model.PredictionResult
import java.time.LocalDate
import java.time.temporal.ChronoUnit
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.pow
import kotlin.math.sqrt

@Singleton
class CyclePredictionEngine @Inject constructor() {

    fun buildPrediction(
        periods: List<Period>,
        referenceDate: LocalDate = LocalDate.now(),
        fallbackCycleLength: Int = 28,
        fallbackPeriodLength: Int = 5,
        lutealPhaseLength: Int = 14
    ): PredictionResult? {
        val normalized = periods
            .sortedByDescending { it.startDate }
            .distinctBy { it.startDate }

        if (normalized.isEmpty()) return null

        val cycleLengths = normalized
            .zipWithNext { current, previous ->
                ChronoUnit.DAYS.between(previous.startDate, current.startDate).toInt()
            }
            .filter { it in 15..90 }

        val averageCycleLength = when {
            cycleLengths.isEmpty() -> fallbackCycleLength
            cycleLengths.size < 3 -> cycleLengths.average().toInt().coerceIn(21, 40)
            else -> weightedCycleAverage(cycleLengths)
        }

        val periodLengths = normalized
            .mapNotNull { record ->
                record.endDate?.let {
                    ChronoUnit.DAYS.between(record.startDate, it).toInt() + 1
                }
            }
            .filter { it in 1..14 }

        val averagePeriodLength = if (periodLengths.isEmpty()) {
            fallbackPeriodLength
        } else {
            periodLengths.average().toInt().coerceIn(2, 10)
        }

        val lastStart = normalized.first().startDate
        val nextPeriodStart = if (lastStart.isAfter(referenceDate)) {
            lastStart
        } else {
            var predicted = lastStart
            while (!predicted.isAfter(referenceDate)) {
                predicted = predicted.plusDays(averageCycleLength.toLong())
            }
            predicted
        }

        val nextPeriodEnd = nextPeriodStart.plusDays((averagePeriodLength - 1).toLong())
        val ovulationDate = nextPeriodStart.minusDays(lutealPhaseLength.toLong())
        val fertileWindowStart = ovulationDate.minusDays(5)
        val fertileWindowEnd = ovulationDate.plusDays(1)

        val stdDeviation = standardDeviation(cycleLengths)
        val variabilityScore = (1f - (stdDeviation / 12f)).coerceIn(0.05f, 1f)
        val sampleScore = (cycleLengths.size / 6f).coerceIn(0.25f, 1f)
        val confidenceScore = (variabilityScore * 0.7f + sampleScore * 0.3f).coerceIn(0.1f, 0.99f)

        return PredictionResult(
            nextPeriodStart = nextPeriodStart,
            nextPeriodEnd = nextPeriodEnd,
            ovulationDate = ovulationDate,
            fertileWindowStart = fertileWindowStart,
            fertileWindowEnd = fertileWindowEnd,
            averageCycleLength = averageCycleLength,
            averagePeriodLength = averagePeriodLength,
            variabilityScore = variabilityScore,
            confidenceScore = confidenceScore,
            isIrregular = stdDeviation >= 4.5f,
            cycleLengthStdDeviation = stdDeviation
        )
    }

    private fun weightedCycleAverage(cycleLengths: List<Int>): Int {
        val recentFirst = cycleLengths
        var weight = recentFirst.size.toFloat()
        var weightedTotal = 0f
        var totalWeight = 0f
        recentFirst.forEach { length ->
            weightedTotal += length * weight
            totalWeight += weight
            weight = (weight - 1f).coerceAtLeast(1f)
        }
        return (weightedTotal / totalWeight).toInt().coerceIn(21, 40)
    }

    private fun standardDeviation(values: List<Int>): Float {
        if (values.isEmpty()) return 0f
        val mean = values.average()
        val variance = values.map { (it - mean).pow(2) }.average()
        return sqrt(variance).toFloat()
    }
}

File: cyclecare_flutter/lib/domain/engines/cycle_prediction_engine.dart
import 'dart:math' as math;
import '../entities/period.dart';
import '../entities/cycle.dart';

class CyclePredictionEngine {
  PredictionResult? buildPrediction({
    required List<Period> periods,
    DateTime? referenceDate,
    int fallbackCycleLength = 28,
    int fallbackPeriodLength = 5,
    int lutealPhaseLength = 14,
  }) {
    final ref = referenceDate ?? DateTime.now();
    
    // Normalize periods
    final normalized = periods
        .toList()
      ..sort((a, b) => b.startDate.compareTo(a.startDate));
    
    // Remove duplicates by start date
    final uniquePeriods = <Period>[];
    final seenDates = <DateTime>{};
    for (final period in normalized) {
      if (!seenDates.contains(period.startDate)) {
        uniquePeriods.add(period);
        seenDates.add(period.startDate);
      }
    }

    if (uniquePeriods.isEmpty) return null;

    // Calculate cycle lengths
    final cycleLengths = <int>[];
    for (int i = 0; i < uniquePeriods.length - 1; i++) {
      final current = uniquePeriods[i];
      final previous = uniquePeriods[i + 1];
      final length = current.startDate.difference(previous.startDate).inDays;
      if (length >= 15 && length <= 90) {
        cycleLengths.add(length);
      }
    }

    // Calculate average cycle length
    final int averageCycleLength;
    if (cycleLengths.isEmpty) {
      averageCycleLength = fallbackCycleLength;
    } else if (cycleLengths.length < 3) {
      averageCycleLength = (cycleLengths.reduce((a, b) => a + b) / cycleLengths.length)
          .round()
          .clamp(21, 40);
    } else {
      averageCycleLength = _weightedCycleAverage(cycleLengths);
    }

    // Calculate period lengths
    final periodLengths = uniquePeriods
        .where((p) => p.endDate != null)
        .map((p) => p.endDate!.difference(p.startDate).inDays + 1)
        .where((length) => length >= 1 && length <= 14)
        .toList();

    final int averagePeriodLength;
    if (periodLengths.isEmpty) {
      averagePeriodLength = fallbackPeriodLength;
    } else {
      averagePeriodLength = (periodLengths.reduce((a, b) => a + b) / periodLengths.length)
          .round()
          .clamp(2, 10);
    }

    // Predict next period start
    final lastStart = uniquePeriods.first.startDate;
    DateTime nextPeriodStart;
    
    if (lastStart.isAfter(ref)) {
      nextPeriodStart = lastStart;
    } else {
      nextPeriodStart = lastStart;
      while (!nextPeriodStart.isAfter(ref)) {
        nextPeriodStart = nextPeriodStart.add(Duration(days: averageCycleLength));
      }
    }

    final nextPeriodEnd = nextPeriodStart.add(Duration(days: averagePeriodLength - 1));
    final ovulationDate = nextPeriodStart.subtract(Duration(days: lutealPhaseLength));
    final fertileWindowStart = ovulationDate.subtract(const Duration(days: 5));
    final fertileWindowEnd = ovulationDate.add(const Duration(days: 1));

    // Calculate confidence
    final stdDeviation = _standardDeviation(cycleLengths);
    final variabilityScore = (1.0 - (stdDeviation / 12.0)).clamp(0.05, 1.0);
    final sampleScore = (cycleLengths.length / 6.0).clamp(0.25, 1.0);
    final confidenceScore = (variabilityScore * 0.7 + sampleScore * 0.3).clamp(0.1, 0.99);

    return PredictionResult(
      nextPeriodStart: nextPeriodStart,
      nextPeriodEnd: nextPeriodEnd,
      ovulationDate: ovulationDate,
      fertileWindowStart: fertileWindowStart,
      fertileWindowEnd: fertileWindowEnd,
      averageCycleLength: averageCycleLength,
      averagePeriodLength: averagePeriodLength,
      variabilityScore: variabilityScore,
      confidenceScore: confidenceScore,
      isIrregular: stdDeviation >= 4.5,
      cycleLengthStdDeviation: stdDeviation,
    );
  }

  int _weightedCycleAverage(List<int> cycleLengths) {
    double weight = cycleLengths.length.toDouble();
    double weightedTotal = 0.0;
    double totalWeight = 0.0;
    
    for (final length in cycleLengths) {
      weightedTotal += length * weight;
      totalWeight += weight;
      weight = (weight - 1.0).clamp(1.0, double.infinity);
    }
    
    return (weightedTotal / totalWeight).round().clamp(21, 40);
  }

  double _standardDeviation(List<int> values) {
    if (values.isEmpty) return 0.0;
    
    final mean = values.reduce((a, b) => a + b) / values.length;
    final variance = values
        .map((v) => math.pow(v - mean, 2))
        .reduce((a, b) => a + b) / values.length;
    
    return math.sqrt(variance);
  }
}
